File: js/packages/models/src/note-delta.ts
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import NoteElementType from './note-element-type';
import NoteRecord from './note-record';
import ValueDelta from './value-delta';

export default class NoteDelta {
  #valueDelta?: ValueDelta;
  #elementType?: NoteElementType;
  #noteTypeID?: string;

  constructor(readonly ID: string) {}

  editValue(d: ValueDelta): NoteDelta {
    this.#valueDelta =
        (this.#valueDelta === void 0) ? d : this.#valueDelta.compose(d);
    return this;
  }
  setElementType(t: NoteElementType): NoteDelta {
    this.#elementType = t;
    return this;
  }
  setNoteType(t: string): NoteDelta {
    this.#noteTypeID = t;
    return this;
  }

  apply(prev: NoteRecord): NoteRecord {
    return {
      ID : prev.ID,
      value : (this.#valueDelta === void 0)
                  ? prev.value
                  : this.#valueDelta.apply(prev.value),
      elementType : (this.#elementType === void 0) ? prev.elementType
                                                   : this.#elementType,
      noteTypeID : (this.#noteTypeID === void 0) ? prev.noteTypeID
                                                 : this.#noteTypeID,
      childrenIDs : prev.childrenIDs,
    };
  }
}

File: dart/nm_delta/lib/src/note_delta.dart
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'item_op.dart';
import 'sequence_delta.dart';
import 'string_delta.dart';

class NoteID extends ItemOp<String> {
  NoteID.put(String item) : super.put(item);
  NoteID.delete() : super.delete();
}

class NoteIDs extends SequenceDelta<String> {
  NoteIDs.insert(Iterable<String> ids) : super.insert(ids);
  NoteIDs.retain(int length) : super.retain(length);
  NoteIDs.delete(int length) : super.delete(length);
  NoteIDs.unmodifiable(Iterable<SequenceOp<String>> ops)
      : super.unmodifiable(ops);
}

class NoteValue extends StringDelta {
  NoteValue.insertString(String text) : super.insertString(text);
  NoteValue.retain(int length) : super.retain(length);
  NoteValue.delete(int length) : super.delete(length);
  NoteValue.unmodifiable(Iterable<SequenceOp<int>> ops)
      : super.unmodifiable(ops);
}

class NoteDelta {
  final StringDelta value;
  final ItemOp<String> valueTypeID;
  final SequenceDelta<String> contentIDs;
  final SequenceDelta<String> typeIDs;
  final bool _isBase;
  final bool _modifiable;

  NoteDelta({this.value, this.valueTypeID, this.contentIDs, this.typeIDs})
      : _isBase = null,
        _modifiable = true;

  NoteDelta.unmodifiable(
      {StringDelta value,
      this.valueTypeID,
      SequenceDelta<String> contentIDs,
      SequenceDelta<String> typeIDs})
      : value = value == null ? null : value.toUnmodifiable(),
        contentIDs = contentIDs == null ? null : contentIDs.toUnmodifiable(),
        typeIDs = typeIDs == null ? null : typeIDs.toUnmodifiable(),
        _isBase = (value == null || value.isBase) &&
            (valueTypeID == null || valueTypeID.isBase) &&
            (contentIDs == null || contentIDs.isBase) &&
            (typeIDs == null || typeIDs.isBase),
        _modifiable = false;

  bool get isBase => _modifiable
      ? (value == null || value.isBase) &&
          (valueTypeID == null || valueTypeID.isBase) &&
          (contentIDs == null || contentIDs.isBase) &&
          (typeIDs == null || typeIDs.isBase)
      : _isBase;

  Map<String, dynamic> toJson() {
    var json = <String, dynamic>{};
    if (value != null) {
      json['v'] = value.toJson();
    }
    if (valueTypeID != null) {
      json['vt'] = valueTypeID.toJson();
    }
    if (contentIDs != null) {
      json['cs'] = contentIDs.toJson();
    }
    if (typeIDs != null) {
      json['ts'] = typeIDs.toJson();
    }
    return json;
  }

  NoteDelta toUnmodifiable() => _modifiable
      ? NoteDelta.unmodifiable(
          value: value,
          valueTypeID: valueTypeID,
          contentIDs: contentIDs,
          typeIDs: typeIDs)
      : this;

  NoteDelta apply(NoteDelta d) {
    return NoteDelta.unmodifiable(
        value: value == null
            ? d.value
            : d.value == null
                ? value
                : value.apply(d.value),
        valueTypeID: valueTypeID == null
            ? d.valueTypeID
            : d.valueTypeID == null
                ? valueTypeID
                : valueTypeID.apply(d.valueTypeID),
        contentIDs: contentIDs == null
            ? d.contentIDs
            : d.contentIDs == null
                ? contentIDs
                : contentIDs.apply(d.contentIDs),
        typeIDs: typeIDs == null
            ? d.typeIDs
            : d.typeIDs == null
                ? typeIDs
                : typeIDs.apply(d.typeIDs));
  }
}
